src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,134 @@
+    constexpr variant(const variant&) = default;
+    constexpr variant(variant&&) = default;
+
+    // Ensure that this is not implicitly convertable from std::variant
+    template<class T>
+    constexpr variant(T&& t) noexcept // NOLINT(*-explicit-*)
+    requires(!std::is_same_v<std::decay_t<T>, std::variant<Types...>>)
+      : type(std::forward<T>(t)){};
+    explicit constexpr variant(std::variant<Types...> v) noexcept
+      : type(std::move(v)){};
+
+    template<class T, class... Args>
+    constexpr explicit variant(std::in_place_type_t<T> in_place, Args&&... args)
+      : type(in_place, std::forward<Args...>(args)...) {}
+    template<std::size_t I, class... Args>
+    constexpr explicit variant(
+      std::in_place_index_t<I> in_place, Args&&... args)
+      : type(in_place, std::forward<Args...>(args)...) {}
+
+    variant& operator=(const variant&) = default;
+    variant& operator=(variant&&) = default;
+
+    constexpr ~variant() = default;
+
+    using type::emplace;
+    using type::index;
+    using type::swap;
+    using type::valueless_by_exception;
+};
+
+template<typename... T>
+void tag_invoke(tag_t<write_tag>, iobuf& out, variant<T...> v) {
+    write<size_t>(out, v.index());
+    std::visit(
+      [&out]<typename V>(V&& v) { write(out, std::forward<V>(v)); },
+      std::move(v));
+}
+
+namespace detail {
+
+template<typename Variant>
+struct variant_discrimiant_factory {
+    using constructor = Variant (*)(iobuf_parser&, std::size_t);
+    using constructor_table
+      = std::array<constructor, std::variant_size_v<Variant>>;
+
+    consteval variant_discrimiant_factory()
+      : constructors([]<std::size_t... Index>(std::index_sequence<Index...>) {
+          return std::to_array<constructor>({
+            [](iobuf_parser& in, std::size_t bytes_left_limit) {
+                return Variant{
+                  std::in_place_index<Index>,
+                  read_nested<std::variant_alternative_t<Index, Variant>>(
+                    in, bytes_left_limit)};
+            }...,
+          });
+      }(std::make_index_sequence<std::variant_size_v<Variant>>())) {}
+
+    constructor_table constructors;
+};
+
+} // namespace detail
+
+template<typename... T>
+void tag_invoke(
+  tag_t<read_tag>,
+  iobuf_parser& in,
+  variant<T...>& t,
+  const std::size_t bytes_left_limit) {
+    using Type = std::decay_t<decltype(t)>;
+    using UnderlyingType = Type::type;
+
+    auto index = read_nested<size_t>(in, bytes_left_limit);
+
+    if (index >= std::variant_size_v<UnderlyingType>) {
+        throw serde_exception(fmt_with_ctx(
+          ssx::sformat,
+          "reading type {} of size {}: {} bytes left - unexpected variant "
+          "index: {}, variant size: {}",
+          type_str<Type>(),
+          sizeof(Type),
+          in.bytes_left(),
+          index,
+          std::variant_size_v<UnderlyingType>));
+    }
+    constexpr static detail::variant_discrimiant_factory<UnderlyingType>
+      factory{};
```

Comment:
The `variant_discrimiant_factory` is `consteval`, is having an instance with static storage duration a pessimization?

Reply:
Probably, TBH I don't know if there is any difference when clang is done optimizing. I've removed static storage.

Reply:
https://gcc.godbolt.org/z/qqaG7Gqxo same codegen